### ISSUE TITLE
gixref 0.41 fails to compile in CI

### ISSUE THREAD

**Issue by alice-i-cecile** [Byron/gitoxide#1309]
> the method `context` exists for fn item `fn(&mut &[u8]) -> Result<SignatureRef<'_>, ErrMode<_>> {decode::<'_, _>}`, but its trait bounds were not satisfied

Spotted in https://github.com/bevyengine/bevy/actions/runs/8041691846/job/21961284606#step:4:409. This sort of spontaneous failure is generally caused by a semver violation somewhere.

**Comment by wasabrot** [comment]
Also cargo install cargo-binstall is failing due to this, I suppose many projects are affected

**Comment by Byron** [comment]
This should be fixed now. I investigated the update issue of `gix-ref` and why it wouldn't pull in the latest version of  it, and it turned out that it was all based on other patch-releases needing `gix-actor` in the latest version, so the resolver did the right thing and found a lower version of `gix-ref` which was compatible with the latest version of `gix-actor`, at least on paper.

Since I wasn't really able to produce a `gix-ref` which can work with both versions of `gix-actor` as it would then have to mix `winnow` 0.5 and 0.6 which doesn't work, I decided to yank `gix` v0.59 entirely as it's the least used.

The culprit here is that an upgrade to `winnow` 0.59 happened without me realizing that this is a breaking change as it's not perfectly isolated. `gix-actor` for instance leaks an error type (for good reason), but that's also causing the issue.

In future, and it's something I usually do, I have to treat breaking `winnow` upgrades as breaking for the API, and even though it's something I kind-of know, this time it slipped by in the reviews. As the upgrade was contributed, it was less obvious to me I suppose, but I have to be more vigilant.

My greatest apologies for the disruption this has caused :/.

**Comment by wasabrot** [comment]
Thx a lot @Byron  such a fast response is really great; right now I m testing with the script which is already working; later I will re-test with what caused the break and give short feedback here of the outcomes.

**Comment by alice-i-cecile** [comment]
Thanks a ton for the fast reply :)

**Comment by leighmcculloch** [comment]
> I investigated the update issue of `gix-ref` and why it wouldn't pull in the latest version of it, and it turned out that it was all based on other patch-releases needing `gix-actor` in the latest version, so the resolver did the right thing and found a lower version of `gix-ref` which was compatible with the latest version of `gix-actor`, at least on paper.

Thanks @Byron for shipping a fix so quickly. For projects that hadn't updated gix or winnow, or changed their dependencies, do you know how the release of gix v0.59 cause builds that had Cargo.lock files present and dependent on gix v0.58 / gix-actor v0.30.0 / gix-object v0.41.0 to pickup the new versions gix v0.59 / gix-actor v0.30.1 / gix-object v0.41.1?

**Comment by NobodyXu** [comment]
Because `cargo-install` does not use the lockfile by default, unless you pass `--locked` to it

**Comment by Byron** [comment]
> Might be worth looking into running cargo-semver in CI: these sort of issues are really tricky to spot manually.

I agree. Previously I was repelled by the anticipated CI cost of this, how much would it be able to do within a 30min window at most? But that was admittedly based mostly on feeling, not at all on data. I also know it can't detect everything yet, and in order to catch this, you'd really have to compile the current version of `gix-actor` with the most recent released version of `gix-ref`. See, I still don't think anything could really catch this in time, but what should work is a CI job that installs the previous version of `gix` which should then have failed just as early.

**Comment by alice-i-cecile** [comment]
@obi1kenobi, the tool author, will have better answers for you here :)

**Comment by obi1kenobi** [comment]
Quick thoughts:
- The extremely customizable tool you describe (at least as I understand it) sounds very doable.
- Using `rustc` itself is absolutely doable too (we have a working proof-of-concept!), and is merely a question of having ongoing funding to keep it supported as `rustc`'s internals change often.
- 100% certainty is impossible because of the Halting Problem (new bugs are arguably breaking changes), but a good test suite + custom lints system we used at my last job ([10min talk I gave on it](https://www.hytradboi.com/2022/how-to-query-almost-everything)) felt like it got us to damn near 100%.

The bottom line is that with a steady source of funding, I think we have a very solid shot at all this — in a way where the entire Rust ecosystem could win, not just `gitoxide` and projects that use it.

I'd love your thoughts! Might you be up for a chat sometime in the next couple of weeks?

**Comment by Byron** [comment]
Yes, and I sent you an email hoping we can schedule that call this week even :D.
